packages/astro/src/vite-plugin-astro/index.ts

```diff
@@ -96,6 +97,25 @@ export default function astro({ config, devServer }: AstroPluginOptions): vite.P
             err.sourceLoc = { file: id, line: (focusMapping[0][2] || 0) + 1, column: (focusMapping[0][3] || 0) + 1 };
           }
         }
+
+        let message = '';
+        if (typeof err === 'string') message = err;
+        if (err && typeof err === 'object') {
+          if (err.stack && typeof err.stack === 'object' && err.stack.message) {
+            message = err.stack.message;
+          } else if (typeof err.toString === 'function') {
+            message = err.toString();
+          }
+        }
+
+        // WASM panic
+        if (message && message === 'RuntimeError: unreachable') {
+          err.message = 'Syntax error';
```

Comment:
Rather than throw “unknown error“ or “unrecoverable error“ I think in almost every case it can be classified as a syntax error: we encountered an unexpected input.

Is this clearer to users to fix?

Reply:
I think syntax error is clear for when it is a syntax error, but I don't think that it would necessarily always be a syntax error, no? Is it reasonable to output "syntax error" when it's a bug in Astro on valid syntax 🤔

Reply:
I would expect a syntax error to be able to point to the line in question that it ran into the problem with, whereas this seems like a general catchall for errors? Or am I wrong? 😅

Reply:
> Is it reasonable to output "syntax error" when it's a bug in Astro on valid syntax 🤔

Yeah good point. I can’t guarantee that we don’t have any of those “panic“s still uncaught. I think this PR is hopeful that we do catch all of the “correct” Astro syntax bugs, and panics are for everything else.

Reply:
> I would expect a syntax error to be able to point to the line in question that it ran into the problem with

Right! We do need to work on that. We just don’t have that available in the compiler, so if an error throws there we can’t give the line/column but it is a syntax error (even if it’s a syntax error for code that _did_ work in 0.20 and we just need to fix in our new compiler).

For more context, the compiler _does_ update the sourcemap, and if an error is thrown in JS execution, Vite can point you exactly to the line/column of the error. But when the error happens in the _compiler_, even though it has the sourcemap, we haven’t built true syntax checking there so it just fails and often doesn’t even know what character was bad until it’s already crashed.

Reply:
huh, ok

i guess this is another case where I'm still not fully up-to-date on how the new layout of astro works!

Reply:
that makes more sense.. in the sense that there's good syntax checking in vite, but only "oops everything blew up" syntax checking in the go compiler atm?

That's definitely something that I would really like to be a part of improving then 😁 as the go compiler becomes stable

Reply:
> in the sense that there's good syntax checking in vite, but only "oops everything blew up" syntax checking in the go compiler atm?

Yeah basically. Another way to put it is the compiler currently expects perfect syntax, and doesn’t properly check for invalid syntax (which it has to do because JS is a foreign language and we can’t just execute it)